obsreport/obsreport_exporter.go

```diff
@@ -0,0 +1,216 @@
+		stats.UnitDimensionless)
+	mExporterFailedToSendSpans = stats.Int64(
+		exporterPrefix+FailedToSendSpansKey,
+		"Number of spans in failed attempts to send to destination.",
+		stats.UnitDimensionless)
+	mExporterSentMetricPoints = stats.Int64(
+		exporterPrefix+SentMetricPointsKey,
+		"Number of metric points successfully sent to destination.",
+		stats.UnitDimensionless)
+	mExporterFailedToSendMetricPoints = stats.Int64(
+		exporterPrefix+RefusedMetricPointsKey,
+		"Number of metric points in failed attempts to send to destination.",
+		stats.UnitDimensionless)
+)
+
+// StartTraceDataExportOp is called at the start of an Export operation.
+// The returned context should be used in other calls to the obsreport functions
+// dealing with the same export operation.
+func StartTraceDataExportOp(ctx context.Context, exporter string) (context.Context, *trace.Span) {
+	return traceExportDataOp(
+		exporterContext(ctx, exporter),
+		exporter,
+		exportTraceDataOperationSuffix)
+}
+
+// EndTraceDataExportOp completes the export operation that was started with
+// StartTraceDataExportOp.
+func EndTraceDataExportOp(
+	ctx context.Context,
+	span *trace.Span,
+	numExportedSpans int,
+	numDroppedSpans int, // TODO: For legacy measurements, to be removed in the future.
+	err error,
+) {
+	if useLegacy {
+		observability.RecordMetricsForTraceExporter(
+			ctx, numExportedSpans, numDroppedSpans)
+	}
+
+	endExportOp(
+		ctx,
+		span,
+		numExportedSpans,
+		err,
+		traceData,
+	)
+}
+
+// StartMetricsExportOp is called at the start of an Export operation.
+// The returned context should be used in other calls to the obsreport functions
+// dealing with the same export operation.
+func StartMetricsExportOp(
+	ctx context.Context,
+	exporter string,
+) (context.Context, *trace.Span) {
+	return traceExportDataOp(
+		exporterContext(ctx, exporter),
+		exporter,
+		exportMetricsOperationSuffix)
+}
+
+// EndMetricsExportOp completes the export operation that was started with
+// StartMetricsExportOp.
+func EndMetricsExportOp(
+	ctx context.Context,
+	span *trace.Span,
+	numExportedPoints int,
+	numExportedTimeSeries int, // TODO: For legacy measurements, to be removed in the future.
+	numDroppedTimeSeries int, // TODO: For legacy measurements, to be removed in the future.
+	err error,
+) {
+	if useLegacy {
+		observability.RecordMetricsForMetricsExporter(
+			ctx, numExportedTimeSeries, numDroppedTimeSeries)
+	}
+
+	endExportOp(
+		ctx,
+		span,
+		numExportedPoints,
+		err,
+		metricsData,
+	)
+}
+
+// exporterContext adds the keys used when recording observability metrics to
+// the given context returning the newly created context. This context should
+// be used in related calls to the obsreport functions so metrics are properly
+// recorded.
+func exporterContext(
+	ctx context.Context,
+	exporter string,
+) context.Context {
+	if useLegacy {
+		ctx = observability.ContextWithExporterName(ctx, exporter)
+	}
+
+	if useNew {
+		ctx, _ = tag.New(ctx, tag.Upsert(
```

Comment:
Does this need to happen every time we begin to send a batch of spans? I am not familiar with metrics API, but I wonder if there is a way to only do these calls once when exporter is created.

Reply:
It could be made as an optimization. Currently, only OC receiver benefits from that - it is a matter of just exposing one of the private functions of the package. That said the initial context needs to be copied anyway for individual operations. I will have to measure with benchmarks to see if it is worth the trouble - will do it when moving the receivers to this new package.